### PR TITLE
[SYCLomatic] Analyze the un-initialized scalar value in sycl group/algorithm APIs

### PR DESCRIPTION
**Pull request by zhiweij1**
1. If the tool can find the declaration, add the initialization.
2. If failed, emit a warning.

Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/Utility.h [hunk 1]
```diff
@@ -581,6 +581,19 @@ class PairedPrinter {
...
+/// @brief Check if an argument is initialized.
+/// @param Arg Function call argument.
+/// @param DeclsNeedToBeInitialized UnInitialized VarDecl(s). Only can be used
+/// when return 0.
+/// @return  1: Initialized
+///          0: Not initialized.
+///         -1: Cannot deduce.
+int isArgumentInitialized(
+    const clang::Expr *Arg,
+    std::vector<const clang::VarDecl *> &DeclsNeedToBeInitialized);
```
Comment (intwanghao) [inline clang/lib/DPCT/Utility.h]: Does this function work for pointer type or array type? If not, should we need to consider above two type?
Like:
__shfl_xor_sync(0xFFFFFFFF, val[index] or *val, laneMask, WS);

File: clang/test/dpct/warp.cu [hunk 2]
```diff
@@ -327,7 +327,7 @@ __global__ void kernel34() {
 }
 
 __global__ void kernel35() {
-  // CHECK: int val;
+  // CHECK: int val = 0;
```
Comment (intwanghao) [inline clang/test/dpct/warp.cu]: May need add test case for new warning message and slightly more complex case like:
int val;
val = ... or func(val);
__shfl_xor_sync(0xFFFFFFFF, val, laneMask, WS);

File: clang/lib/DPCT/Utility.h (hunk 1 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.h]: This issue only occurs for scalar type parameters.

File: clang/test/dpct/warp.cu (hunk 2 above)
Comment (zhiweij1) [reply clang/test/dpct/warp.cu]: Will do

Comment (zhiweij1) [reply clang/test/dpct/warp.cu]: done

**Review by intwanghao — APPROVED** [review]
LGTM

File: clang/lib/DPCT/ASTTraversal.h [hunk 3]
```diff
@@ -259,6 +259,26 @@ template <typename T> class NamedMigrationRule : public MigrationRule {
...
+      return;
+    std::vector<const clang::VarDecl *> DeclsNeedToBeInitialized;
+    int Res = isArgumentInitialized(Arg, DeclsNeedToBeInitialized);
+    if (Res == 0) {
+      for (const auto D : DeclsNeedToBeInitialized) {
+        emplaceTransformation(new InsertText(
+            D->getEndLoc().getLocWithOffset(Lexer::MeasureTokenLength(
+                D->getEndLoc(), DpctGlobalInfo::getSourceManager(),
+                DpctGlobalInfo::getContext().getLangOpts())),
+            " = 0"));
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ASTTraversal.h:274]: If there have the same issue with vector types, should we use zero initialization?

**Review by yihwang-nv — APPROVED** [review]
LGTM

File: clang/lib/DPCT/ASTTraversal.h (hunk 3 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.h:274]: vector types have default ctor to initialize their members.

File: clang/test/dpct/warp.cu [hunk 4]
```diff
@@ -337,6 +337,36 @@ __global__ void kernel35() {
   __shfl_xor_sync(0xFFFFFFFF, val, laneMask, WS);
 }
 
+__device__ void device36(int val, int laneMask, int WS) {
+  // CHECK: /*
+  // CHECK-NEXT: DPCT1121:{{[0-9]+}}: Please initialize "val" which is used in the SYCL group function/algorithm.
```
Comment (zhimingwang36) [inline clang/test/dpct/warp.cu]:
```suggestion
  // CHECK-NEXT: DPCT1121:{{[0-9]+}}: Make sure "val" which is used in the SYCL group function/algorithm be initialized. 
```

File: clang/test/dpct/warp.cu [hunk 5]
```diff
@@ -337,6 +337,36 @@ __global__ void kernel35() {
...
+  // CHECK-NEXT: */
+  // CHECK-NEXT: dpct::permute_sub_group_by_xor(item_ct1.get_sub_group(), val, laneMask, WS);
+  __shfl_xor_sync(0xFFFFFFFF, val, laneMask, WS);
+}
+
+__global__ void kernel36() {
+  int val;
+  int laneMask;
+  int WS;
+  device36(val, laneMask, WS);
```
Comment (zhimingwang36) [inline clang/test/dpct/warp.cu]: please test if we need to add following two cases. 
 device36(val + 10, laneMak, WS)
int b=0;
 device36(val + b , laneMak, WS)

File: clang/lib/DPCT/Utility.h [hunk 6]
```diff
@@ -581,6 +581,19 @@ class PairedPrinter {
   ~PairedPrinter() { OS << Postfix; }
 };
 std::string appendPath(const std::string &P1, const std::string &P2);
+std::set<const clang::DeclRefExpr *>
+matchTargetDREInScope(const clang::VarDecl *TargetDecl,
+                      const clang::Stmt *Range);
+/// @brief Check if an argument is initialized.
+/// @param Arg Function call argument.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Utility.h]:
```suggestion
/// @param Arg Function call argument. (may be a expr)
```

File: clang/lib/DPCT/CallExprRewriter.h [hunk 7]
```diff
@@ -259,6 +259,45 @@ class ReportWarningRewriterFactory
...
+            D->getEndLoc().getLocWithOffset(Lexer::MeasureTokenLength(
+                D->getEndLoc(), DpctGlobalInfo::getSourceManager(),
+                DpctGlobalInfo::getContext().getLangOpts()));
+        std::string Repl = " = 0";
+        DpctGlobalInfo::getInstance().addReplacement(
+            std::make_shared<ExtReplacement>(DpctGlobalInfo::getSourceManager(),
+                                             InsertLoc, 0, Repl, nullptr));
+      }
+    } else if (Res == -1) {
+      BaseT::create(C);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CallExprRewriter.h:296]: add comment for else if  and L297.

File: clang/lib/DPCT/CallExprRewriter.h [hunk 8]
```diff
@@ -259,6 +259,45 @@ class ReportWarningRewriterFactory
...
 
+class AnalyzeUninitializedDeviceVarRewriterFactory
+    : public CallExprRewriterFactory<
+          UnsupportFunctionRewriter<
+              std::function<const Expr *(const CallExpr *C)>>,
+          Diagnostics, std::function<const Expr *(const CallExpr *C)>> {
+  using BaseT = CallExprRewriterFactory<
+      UnsupportFunctionRewriter<std::function<const Expr *(const CallExpr *C)>>,
+      Diagnostics, std::function<const Expr *(const CallExpr *C)>>;
+  std::shared_ptr<CallExprRewriterFactoryBase> First;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CallExprRewriter.h]:
```suggestion
  std::shared_ptr<CallExprRewriterFactoryBase> First; // need a good name for it.
```

File: clang/lib/DPCT/ASTTraversal.h [hunk 9]
```diff
@@ -259,6 +259,26 @@ template <typename T> class NamedMigrationRule : public MigrationRule {
     emplaceTransformation(std::move(SIT));
   }
 
+  void analyzeUninitializedDeviceVar(const clang::Expr *Call,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.h:262]: add comment for this util function in L262.

File: clang/lib/DPCT/APINamesWarp.inc [hunk 10]
```diff
@@ -9,79 +9,94 @@
...
+                     makeCombinedArg(
+                         makeCallArgCreatorWithCall(0),
+                         makeCombinedArg(
+                             makeCombinedArg(
+                                 ARG(" & (0x1 << "),
+                                 makeCombinedArg(
+                                     SUBGROUP,
+                                     ARG(".get_local_linear_id())) || "))),
+                             makeCallArgCreatorWithCall(1)))))),
+        1))
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APINamesWarp.inc]:
```suggestion
        1)) /*need add comment for the magic numbe 1*/
```

File: clang/test/dpct/warp.cu (hunk 4 above)
Comment (zhiweij1) [reply clang/test/dpct/warp.cu]: done

File: clang/lib/DPCT/CallExprRewriter.h (hunk 8 above)
Comment (zhiweij1) [reply clang/lib/DPCT/CallExprRewriter.h]: done

File: clang/lib/DPCT/APINamesWarp.inc (hunk 10 above)
Comment (zhiweij1) [reply clang/lib/DPCT/APINamesWarp.inc]: I will add this argument before the inner factory.

File: clang/lib/DPCT/Utility.h (hunk 6 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.h]: done

File: clang/lib/DPCT/CallExprRewriter.h (hunk 7 above)
Comment (zhiweij1) [reply clang/lib/DPCT/CallExprRewriter.h:296]: done

File: clang/lib/DPCT/ASTTraversal.h (hunk 9 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.h:262]: done

File: clang/test/dpct/warp.cu (hunk 5 above)
Comment (zhiweij1) [reply clang/test/dpct/warp.cu]: done

File: clang/test/dpct/warp.cu [hunk 11]
```diff
@@ -337,6 +337,65 @@ __global__ void kernel35() {
...
+  // CHECK: int a = 0;
+  // CHECK-NEXT: int b = 0;
+  // CHECK-NEXT: int laneMask;
+  // CHECK-NEXT: int WS;
+  // CHECK-NEXT: dpct::permute_sub_group_by_xor(item_{{[0-9a-z]+}}.get_sub_group(), a + b, laneMask, WS);
+  int a = 0;
+  int b = 0;
+  int laneMask;
+  int WS;
+  __shfl_xor_sync(0xFFFFFFFF, a + b, laneMask, WS);
```
Comment (zhimingwang36) [inline clang/test/dpct/warp.cu]: please remove this kernel39.

Comment (zhiweij1) [reply clang/test/dpct/warp.cu]: Updated.

File: clang/test/dpct/warp.cu [hunk 12]
```diff
@@ -337,6 +337,70 @@ __global__ void kernel35() {
...
+  // CHECK-NEXT: int src;
+  // CHECK-NEXT: dpct::select_from_sub_group(item_{{[0-9a-z]+}}.get_sub_group(), val, src);
+  int val = device37();
+  int src;
+  __shfl(val, src, 32);
+}
+
+__device__ void device38(int a, int b, int src) {
+  // CHECK: /*
+  // CHECK-NEXT: DPCT1121:{{[0-9]+}}: Make sure that the "a + b" which is used in the SYCL group function/algorithm is initialized.
```
Comment (zhimingwang36) [inline clang/test/dpct/warp.cu]:
```suggestion
  // CHECK-NEXT: DPCT1121:{{[0-9]+}}: Make sure that the variable(s) in "a + b" which is used in the SYCL group function/algorithm is initialized.
```

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**